src/test/java/com/proofpoint/discovery/TestIpHostnameAuthManager.java

```diff
@@ -0,0 +1,250 @@
+package com.proofpoint.discovery;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.proofpoint.discovery.client.ServiceDescriptor;
+import com.proofpoint.discovery.client.ServiceState;
+import com.proofpoint.discovery.client.testing.StaticServiceSelector;
+import com.proofpoint.discovery.store.StoreConfig;
+import com.proofpoint.node.NodeInfo;
+import com.proofpoint.testing.SerialScheduledExecutorService;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.ForbiddenException;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+public class TestIpHostnameAuthManager
+{
+    private InMemoryDynamicStore dynamicStore;
+    private final StaticServiceSelector selector = new StaticServiceSelector(new ServiceDescriptor(
+            UUID.randomUUID(),
+            UUID.randomUUID().toString(),
+            "discovery",
+            "general",
+            "/location",
+            ServiceState.RUNNING,
+            ImmutableMap.of("http", "http://localhost:4111")));
+    private final NodeInfo nodeInfo = new NodeInfo("test_environment");
+    private SerialScheduledExecutorService executor;
+
+    @BeforeMethod
+    public void beforeMethod()
+    {
+        initMocks(this);
+        dynamicStore = new InMemoryDynamicStore(new DiscoveryConfig(), new TestingTimeSupplier());
+        executor = new SerialScheduledExecutorService();
+    }
+
+    @Test
+    public void testReplicate()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, selector, nodeInfo, new StoreConfig(), executor);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("127.0.0.1");
+        assertThatCode(() -> authManager.checkAuthReplicate(request)).doesNotThrowAnyException();
+    }
+
+    @Test
+    public void testReplicateMismatch()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, selector, nodeInfo, new StoreConfig(), executor);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("10.20.30.40");
+        assertThatExceptionOfType(ForbiddenException.class)
+                .isThrownBy(() -> authManager.checkAuthReplicate(request))
+                .withMessageContaining("HTTP 403")
+                .withNoCause();
+    }
```

Comment:
Need to test both of these after adding and deleting the requesting discovery server (and advancing `executor`).